feat(examples): Add "Time locked account" account abstraction example

TheMrAI opened the pull request:
# Description of change

Implement "Time locked account" example for abstract accounts.
This example explores the possibility of utilizing the authentication logic.

Each authenticator type can have a number of "reserved" dynamic fields and provide basic functions for attaching, detaching, rotating, checking and borrowing that type.
Properly using these functions is the responsibility of the implementer, but given the language restrictions this is a very flexible pattern.

## Links to any relevant issues

fixes #8534 

## How the change has been tested

- [x] Basic tests (linting, compilation, formatting, unit/integration tests)
- [x] Patch-specific tests (correctness, functionality coverage)
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have checked that new and existing unit tests pass locally with my changes

TheMrAI committed:
feat(examples): Add "Time locked account" account abstraction example

TheMrAI commented on examples/move/account_template/sources/account_template.move:
```diff
@@ -0,0 +1,251 @@
+    self: &mut IOTAccountBuilder,
+    authenticator: Authenticator,
+) {
+    dynamic_field::add(&mut self.account.id, account::authenticator_df_name(), authenticator);
```
If we refactor `account::authenticator_df_name()` to return a `public struct AuthenticatorKey has copy, drop, store {}` instead, this whole function can me removed and the interface simplified further.

TheMrAI commented on examples/move/account_template/sources/account_template.move:
```diff
@@ -0,0 +1,251 @@
+///
+/// Only the account itself can call this function and the dynamic field must refer be a
+/// reserved one.
+public fun rotate_reserved<Name: copy + drop + store, Value: drop + store>(
```
I am a little on the fence about this function, because one may want to rotate all of them at once, in which case these repeated checks are a waste of cycle, but not sure how that could be extracted.

Dkwcs replied:
Why is it needed?

TheMrAI replied:
The account_template provides safety for not accidentally modifying the reserved dynamic fields.
But in practice this is problematic if you would want to rotate your public keys or any data that you need for authentication.
So you need a way to update them somehow, to give some flexibility to account handling.
Now this isn't perfect, but I didn't have a better idea. 

The pattern is still breakable if one rotates out the reserved fields to something else, but I see no real way of properly restricting it in a flexible manner.

TheMrAI commented on examples/move/account_template/tests/account_template_tests.move:
```diff
@@ -0,0 +1,121 @@
+
+    test_scenario::end(scenario_val);
+}
+
```
Not testing any functionality that is tested in the `iota_account`examples.
If this abstraction is acceptable, in a later PR we can refactor the examples appropriately.

TheMrAI commented on examples/move/time_locked/sources/unlock_time.move:
```diff
@@ -0,0 +1,83 @@
+        EEd25519VerificationFailed,
+    );
+}
+
```
With the abstraction of `account_template`, the user will not have to test basic/common functionality when implementing an account.

TheMrAI commented on examples/move/time_locked/sources/time_locked.move:
```diff
@@ -0,0 +1,83 @@
+    account_template::ensure_tx_sender_is_account(self, ctx);
+
+    authenticate_time(self, clock, auth_ctx, ctx);
+    authenticate_ed25519(self, signature, auth_ctx, ctx);
```
Ideally I would want to use the implementation in the `iota_account` example, but didn't want to touch it as long as we are not sure of this approach.
If this is okay, then in a later PR this can be cleaned up.

In any case it shows in however limited capacity that we can kinda build authenticate functionality on top of each other.

TheMrAI commented on examples/move/account_template/sources/account_template.move:
```diff
@@ -0,0 +1,251 @@
+// The following template provides the basic tools for a developer for implementing an abstract account,
+// with the necessary authentication functions, error codes and expectations.
+
+#[error(code = 0)]
```
Just a reminder here if you know by any chance if this could be a problem.
Not sure if there could be an issue with error codes colliding or becoming very unwieldy to the developer when making an account implementation.
For example they use **account_template**, **some_authenticator** in their **best_account** package.
All there having error codes at 0, 1, 2, 3 etc.
Are you aware this possibly causing an issue somewhere?

miker83z replied:
I think we can consider these codes only relative to the module itself. Check this in the sui framework: https://github.com/MystenLabs/sui/blob/0b1a1d351aeb441cea7b1717c725c97dc0d1df6e/crates/sui-framework/packages/sui-framework/sources/transfer.move#L44

Dkwcs commented on examples/move/time_locked/tests/time_locked_tests.move:
```diff
@@ -0,0 +1,165 @@
+fun account_unlocked() {
+    let mut scenario_val = test_scenario::begin(@0x0);
+    let scenario = &mut scenario_val;
+    let public_key = x"cc62332e34bb2d5cd69f60efbb2a36cb916c7eb458301ea36636c4dbb012bd88";
```
Where did you get those parameters like pub_key, digest and signature?
Could you add a comment or something that explains it in short manner ?

TheMrAI replied:
To be honest I stole it from @valeriyr  :D

Dkwcs commented on examples/move/account_template/sources/account_template.move:
```diff
@@ -0,0 +1,251 @@
+    b"An account must have at least one restricted dynamic field set.";
+#[error(code = 2)]
+const EReservedDynamicFieldsListCannotBeSet: vector<u8> =
+    b"'ReservedDfNames' field cannot be set by the user.";
```
We already discussed it with Davo, just leaving here a reminder 
The error is pretty clear, however we can add a small clarification like why we need this

Dkwcs commented on examples/move/account_template/tests/account_template_tests.move:
```diff
@@ -0,0 +1,121 @@
+    // Any field value can be set as a reserved, and for the purposes of this test
+    // the exact value doesn't matter.
+    let reserved_df_example = ReservedDfExample {};
+    builder.add_reserved_field(reserved_df_example, 6);
```
Why it’s not failing, while it does

```
    builder.add_reserved_field(
        account_template::get_reserved_df_names(),
        vector<std::type_name::TypeName>[],
    );
```

TheMrAI replied:
Not sure I follow. Could you elaborate?
Maybe you are referring to the
`get_reserved_df_names()` => `ReservedDfNames`
And that here the example uses `ReservedDfExample` ?

Dkwcs replied:
Oh gosh, yeah, that’s what I was asking, thanks

Dkwcs commented on examples/move/account_template/sources/account_template.move:
```diff
@@ -0,0 +1,251 @@
+#[error(code = 3)]
+const ETransactionSenderIsNotTheAccount: vector<u8> = b"Transaction must be signed by the account.";
+/// It should be emitted every time when an attempt at modifying a restricted dynamic field was made in
+/// an inappropriate scope scope. This scope will generally be defined by the account implementers.
```
I guess 1 scope is extra

Dkwcs commented on examples/move/time_locked/sources/time_locked.move:
```diff
@@ -0,0 +1,83 @@
+///
+/// Looks like an `authenticate` function, but it isn't as it is private. Nor would it provide
+/// satisfactory access protection for the account itself.
+fun authenticate_time(self: &IOTAccount, clock: &Clock, _auth_ctx: &AuthContext, _ctx: &TxContext) {
```
Is it usable only in authenticate function? 
I mean I’m just trying to get the main purpose - so we can authenticate only if amount of time we set is passed?

TheMrAI replied:
It is like a gate that only opens after a certain point in time.
The unlock_time is a point in time, not an amount of time.
So this opens once and it will stay open.

Yep, this function is only for authentication purposes. Anything that receives an `AuthContext` can only be called in authentication scenarios/within authenticate functions, because `AuthContext` can't be made, only by the system (at least this is how I think the design stands).

Dkwcs commented on examples/move/account_template/sources/account_template.move:
```diff
@@ -0,0 +1,251 @@
+    let reserved_found =
+        reserved_df_names.any!(|reserved| reserved == std::type_name::get<Name>()) || is_authenticate(name);
+
+    if (has_to_be_reserved) {
```
It’s a bit tricky for me
Do we have some documentation with the requirements on confluence that explain this flow?

TheMrAI replied:
Nope, as far as I know there is none, only this pseudocode: https://iotafoundation.atlassian.net/wiki/spaces/LSC/pages/2479718716/Time+Locked+Account

But the patterns used here come from: https://github.com/iotaledger/iota/tree/95fa86891d118a8e65f5265a88bef510233171cf/examples/move/iota_account

valeriyr commented on examples/move/account_template/sources/account_template.move:
```diff
@@ -0,0 +1,251 @@
+/// adding any dynamic fields that are not vital for the specified authenticator should be avoided.
+public struct IOTAccountBuilder {
+    account: IOTAccount,
+    reserved_fields: vector<std::type_name::TypeName>,
```
As we discussed already, using just `std::type_name::TypeName` is not enough to define a dynamic field name.
We should use the bytes representation as well, or implement a new utility function in the `dynamic_field` module to calculate the name value by type.

TheMrAI replied:
Hehe I just commented it above :D

TheMrAI replied:
Fixed in: 73999ed61195ac8d64b388646f77dc30fc6d3943

valeriyr commented on examples/move/account_template/sources/account_template.move:
```diff
@@ -0,0 +1,251 @@
+}
+
+/// Construct an IOTAccountBuilder.
+public fun init_account_builder(ctx: &mut TxContext): IOTAccountBuilder {
```
we can rename this function to `builder` or `create_builder`, it will be called like `account_template::builder` which makes it clear that it is related to `account_template`.

TheMrAI replied:
Fixed in: 2e4a0b426a

valeriyr commented on examples/move/account_template/sources/account_template.move:
```diff
@@ -0,0 +1,251 @@
+///
+/// The `AuthenticatorInfo` will be attached as a dynamic field with key provided by: `account::authenticator_df_name()`
+/// This function must be called, otherwise the builder will fail.
+public fun set_authenticator<Authenticator: copy + drop + store>(
+    self: &mut IOTAccountBuilder,
+    authenticator: Authenticator,
+) {
```
I propose to refactor all the builder functions to return the vuilder instance:
```
public fun set_authenticator<Authenticator: copy + drop + store>(
    self: IOTAccountBuilder,
    authenticator: Authenticator,
): IOTAccountBuilder { ... }
```

TheMrAI replied:
Fixed in: 2e4a0b426a

valeriyr commented on examples/move/account_template/sources/account_template.move:
```diff
@@ -0,0 +1,251 @@
+    self: &mut IOTAccountBuilder,
+    authenticator: Authenticator,
+) {
+    dynamic_field::add(&mut self.account.id, account::authenticator_df_name(), authenticator);
```
As I see, we can treat the authenticator as a reserved field, so we can just call `add_reserved_field` here.

TheMrAI replied:
Not yet, need to refactor authenticator_df_name first, but that is the idea. First I will fix everything in this PR, then when splitting do that refactor as well.

TheMrAI replied:
Fixed in: 73999ed61195ac8d64b388646f77dc30fc6d3943

valeriyr commented on examples/move/account_template/sources/account_template.move:
```diff
@@ -0,0 +1,251 @@
+}
+
+/// Attach a `Value` as a dynamic field to the builder.
+public fun add_reserved_field<Name: copy + drop + store, Value: store>(
```
We need to add a function that allows tus o add non-reserved dynamic fields. It is necessary because dynamic fields can be added only by the account itself that is impossible while creating.

TheMrAI replied:
Closing it, before the re-review.

TheMrAI commented on examples/move/account_template/sources/account_template.move:
```diff
@@ -0,0 +1,251 @@
+        ReservedDfNames {},
+    );
+    let reserved_found =
+        reserved_df_names.any!(|reserved| reserved == std::type_name::get<Name>()) || is_authenticate(name);
```
Don't forget to also check the actual value in the Name. The type is not enough in case built-in/framework types are used like a String.

valeriyr commented on examples/move/account_template/sources/account_template.move:
```diff
@@ -0,0 +1,251 @@
+public fun finish_and_share(self: IOTAccountBuilder) {
+    let IOTAccountBuilder { mut account, reserved_fields } = self;
+
+    let authenticator_set = dynamic_field::exists_(&account.id, account::authenticator_df_name());
+    assert!(authenticator_set, EAuthenticatorNotSet);
```
If we want a user force to set the authenticator, we can pass it to the `build` function and remove `set_authenticator`.

TheMrAI replied:
Fixed in: 2e4a0b426a

valeriyr commented on examples/move/account_template/sources/account_template.move:
```diff
@@ -0,0 +1,251 @@
+/// Finish building the `IOTAccount` and share the object.
+///
+/// The call fails if no authenticator or reserved dynamic fields were set.
+public fun finish_and_share(self: IOTAccountBuilder) {
```
I propose returning the IOTAccount here without sharing and adding the following function:
```rust
public fun share(self: IOTAccount) {
    iota::transfer::share_object(account);
}
```
It makes it possible to do something with the account instance before sharing, for example, get the address.

TheMrAI replied:
Fixed in: 2e4a0b426a

valeriyr commented on examples/move/account_template/sources/account_template.move:
```diff
@@ -0,0 +1,251 @@
+    let authenticator_set = dynamic_field::exists_(&account.id, account::authenticator_df_name());
+    assert!(authenticator_set, EAuthenticatorNotSet);
+
+    assert!(!reserved_fields.is_empty(), EReservedDynamicFieldsNotSet);
+    dynamic_field::add(&mut account.id, ReservedDfNames {}, reserved_fields);
```
It looks to me that this collection can be empty; we can add only the authenticator here.
It is useful when you don't need any data for the authenticator.

TheMrAI replied:
You are right. I just couldn't conceive of such a scenario, so far so opted to keep it requiring at least one.

TheMrAI replied:
Fixed in: 73999ed61195ac8d64b388646f77dc30fc6d3943

valeriyr commented on examples/move/account_template/sources/account_template.move:
```diff
@@ -0,0 +1,251 @@
+    ctx: &TxContext,
+) {
+    ensure_tx_sender_is_account(self, ctx);
+    check_df<Name>(self, &name, true);
```
Instead of having a bool param, we can add two functions, one checks that the name is reserved, and another that it is not reserved. It will be easier to read.

TheMrAI replied:
Closing it, before the re-review.

valeriyr commented on examples/move/account_template/sources/account_template.move:
```diff
@@ -0,0 +1,251 @@
+    check_df<Name>(self, &name, true);
+
+    let account_id = &mut self.id;
+    dynamic_field::remove<_, Value>(account_id, name);
+    dynamic_field::add(account_id, name, value);
```
We can add an utility function like this:
```
/// Updates a dynamic field value and returns the previous one.
/// It is supposed that the dynamic field with the given name already exists.
fun update_dynamic_field<Name: copy + drop + store, Value: store>(
    account_id: &mut UID,
    name: Name,
    value: Value
): Value {
    let previous_value = dynamic_field::remove(account_id, name);
    dynamic_field::add(account_id, name, value);
    previous_value
}
```
It makes code cleaner; it is not necessary to explicitly set the value type.

TheMrAI replied:
I am hesitant with such a change. Until we don't need to update it often replacing two lines for 10 just adds noise.
There is exactly one call site at the moment.
When there is more, we can refactor it.

TheMrAI replied:
After our offline discussion I will update it.

TheMrAI replied:
Closing it, before the re-review.

valeriyr commented on examples/move/account_template/sources/account_template.move:
```diff
@@ -0,0 +1,251 @@
+        ReservedDfNames {},
+    );
+    let reserved_found =
+        reserved_df_names.any!(|reserved| reserved == std::type_name::get<Name>()) || is_authenticate(name);
```
if treat the authenticator field as reserved, we don't need to check `is_authenticate(name)` here.

TheMrAI replied:
Good point! This was some legacy oversight.

TheMrAI replied:
Closing it, before the re-review.

valeriyr reviewed (commented):
I personally would prefer to create a dedicated PR for `account_template`, because it is an independent feature from the `time_locked` example.

valeriyr commented:
@TheMrAI I see that the `account_template` file contains `IOTAccount` implementation. Are you going to refactor the original `IOTAccount`? I think it would be better to refactor it and not add a new file.

Dkwcs commented on examples/move/time_locked/sources/time_locked.move:
```diff
@@ -0,0 +1,83 @@
+module time_locked::time_locked;
```
I guess that is missing 
// Copyright (c) 2025 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

TheMrAI replied:
Dang it. I despise those copyright things XD

TheMrAI replied:
Fixed in: 2e4a0b426a

valeriyr commented on examples/move/iotaccount/sources/basic_keyed_account.move:
```diff
@@ -133,3 +138,45 @@ public fun borrow_public_key(account: &IOTAccount): &vector<u8> {
+public fun authenticate_ed25519_signature(
+    account_id: &UID,
+    signature: vector<u8>,
+    digest: &vector<u8>,
```
This field can be called `message` because, in this case, it reflects the `ed25519_verify` function parameter.
It is not an obligation to sign the digest.

TheMrAI replied:
Fixed in: 0c89d589e2

valeriyr commented on examples/move/iotaccount/sources/basic_keyed_account.move:
```diff
@@ -19,6 +20,10 @@ const EEd25519VerificationFailed: vector<u8> = b"Ed25519 authenticator verificat
+#[error(code = 4)]
+const EPublicKeyAttached: vector<u8> = b"Public key already attached.";
+#[error(code = 5)]
+const EPublicKeyMissing: vector<u8> = b"Public key missing.";
```
3 and 4?

TheMrAI replied:
Fixed in: 0c89d589e2

valeriyr commented on examples/move/iotaccount/sources/iotaccount.move:
```diff
@@ -188,6 +188,10 @@ public fun ensure_tx_sender_is_account(self: &IOTAccount, ctx: &TxContext) {
     assert!(self.id.uid_to_address() == ctx.sender(), ETransactionSenderIsNotTheAccount);
 }
 
+public fun ensure_tx_sender_is_account_id(account_address: &address, ctx: &TxContext) {
```
I would propose to pass `UID` here instead of `address`. `UID` identifies that you call this function from a protected part of code, while `address` can be passed from everywhere.

TheMrAI replied:
Fixed in: 0c89d589e2

valeriyr commented on examples/move/time_locked/sources/unlock_time.move:
```diff
@@ -0,0 +1,122 @@
+    authenticate_unlock_time(id, now);
+}
+
+public fun authenticate_unlock_time(account_id: &UID, current_time: u64) {
```
We can consider adding the following two functions:
```rust
public fun authenticate_with_epoch_timestamp(account_id: &UID, ctx: &TxContext) { ... }
public fun authenticate_with_clock(account_id: &UID, clock: &Clock) { ... }
```
They add protection from passing a wrong value.

TheMrAI replied:
Fixed in: 65c5767dd8

valeriyr commented on examples/move/time_locked/sources/time_locked.move:
```diff
@@ -0,0 +1,122 @@
+
+/// Authenticate access for the `Time locked account`.
+public fun authenticate(
+    id: &UID,
```
This parameter must be changed to `account: & TimeLocked` because it is the account authorization entry point.

TheMrAI replied:
Fixed in: 0c89d589e2

valeriyr commented on examples/move/time_locked/sources/time_locked.move:
```diff
@@ -0,0 +1,122 @@
+
+public fun borrow_uid(self: &TimeLocked): &UID {
+    &self.id
+}
```
This function should be removed.
All the functions you expected to be called from `TimeLocked` need to be explicitly exposed to be gated with the following parameter `self: &TimeLocked`.
Even adding it as `test_only` seems unnecessary.
If you want to test, for example, `authenticate_unlock_time`, you can create a dedicated struct for the test module(TestAccount) and test on it. An example can be found [here](https://github.com/iotaledger/iota/blob/a04b77022f5dd29f1878c9bdec5160a45f6286fe/crates/iota-framework/packages/iota-framework/tests/account_tests.move#L13).

TheMrAI replied:
Fixed in: 17a5ce07fc

valeriyr commented on examples/move/time_locked/sources/utils.move:
```diff
@@ -0,0 +1,122 @@
+    let prev_unlock_time = dynamic_field::remove(account_id, UnlockTime {});
+    dynamic_field::add(account_id, UnlockTime {}, unlock_time);
+    prev_unlock_time
+}
```
If we consider `UnlockTime` as an extension that can be used by other account types, we need to extract it in a dedicated module.
If we don't, we need to change this first parameter to `self: &TimeLocked`. `TimeLocked` shouldn't expose `UID` to users.

TheMrAI replied:
Fixed in: 67be434a92

TheMrAI committed:
Fix review comments part 1

borrow_uid remains until restructuring

TheMrAI committed:
Separeate account from utility

The utility functions for building an account fulfill a completely
different purpose from an actual account implementation.
For this reason it makes sense to separate them as it makes it very
clear what functionality may be problematic from a design perspective
as well.

TheMrAI committed:
Remove borrow_uid from the account

It muddies the interface, especially in terms of access control.
Utility functions aren't concerned with access control, but all
account functions should ensure that all access is correct.
While in this case it doesn't change much, in more granular access
controls it might become important if a clean pattern is established.

TheMrAI committed:
Add convenience functions for unlock utils

TheMrAI committed:
Move utility functions from iotaccount to generic_keyed_authentication

We named the module according to the name of the dynamic field struct.
This is necessary to avoid name collisions within the testing framework
while trying to compose multiple packages.
package_A::name
package_B::name
package_C including both will collide, no matter if an alias or
fully qualified names are used. In package_C one cannot override
the address of pakcage_A or package_B either, so there is no
convenient way to resolve collision.

TheMrAI committed:
Rename generic_keyed_account to keyed_iotaccount

TheMrAI committed:
Rename time_locked::utils to time_locked::unlock_time

TheMrAI committed:
Cover with tests owner_public_key.move

These tests as the module are partial code duplications, but we don't want to
refactor IOTAccount every time some new pattern should emerge.
With this following refactor only has to remove the relevant code and tests
from IOTAccount and transition to using these functions.

valeriyr commented on examples/move/time_locked/sources/account.move:
```diff
@@ -0,0 +1,81 @@
+
+    owner_public_key::authenticate_ed25519(&account.id, signature, ctx.digest());
+    let now = clock.timestamp_ms();
+    unlock_time::authenticate_unlock_time(&account.id, now);
```
Let's use a new function `authenticate_with_clock` here.

TheMrAI replied:
Fixed in: 3d2d1da8bc

valeriyr commented on examples/move/time_locked/sources/unlock_time.move:
```diff
@@ -0,0 +1,82 @@
+    let prev_unlock_time = dynamic_field::remove(account_id, UnlockTime {});
+    dynamic_field::add(account_id, UnlockTime {}, unlock_time);
+    prev_unlock_time
+}
+
+public fun authenticate_with_epoch_timestamp(account_id: &UID, ctx: &TxContext) {
+    authenticate_unlock_time(account_id, ctx.epoch_timestamp_ms())
+}
+
+public fun authenticate_with_clock(account_id: &UID, clock: &Clock) {
+    authenticate_unlock_time(account_id, clock.timestamp_ms())
+}
```
It would be nice to add some tests for these functions.

TheMrAI replied:
I was considering it originally, but it seems a bit excessive as both call `authenticate_unlock_time` directly which is fully tested.
Wdyt?

TheMrAI replied:
Fixed in: 00d607e124

valeriyr reviewed (approved):
I left a few comments, but overall it looks good to me!

TheMrAI committed:
Add documentation to the code.

TheMrAI committed:
Use authenticate_with_clock

TheMrAI committed:
Reorder owner_public_key to follow the move pattern

TheMrAI committed:
Unlock_time additional tests

miker83z reviewed (approved):
LGTM!
